ts-types/ensure.d.ts

```diff
@@ -0,0 +1,15 @@
+export type EnsureFunction = (...args: any[]) => any;
+
+export interface EnsureOptions {
+	name?: string;
+	isOptional?: boolean;
+	default?: any;
+	errorMessage?: string;
+	errorCode?: number;
+	Error?: ErrorConstructor;
+}
+
+type ValidationDatum = [argumentName: string, inputValue: any, ensureFunction: EnsureFunction, options?: object];
+
+declare function ensure(validationDatum: ValidationDatum, ...values: any[]): any[];
```

Comment:
Technically it's `...validationDatum: ValidationDatum, options?: EnsureOptions`. Still I understand it's not possible to define types like that In TS?

Reply:
i need to investigate at this one. I think we may use a tuple but will require multiple declaration i suspect.

```
declare function ensure(validationDatum: ValidationDatum, ...validationDatum: ValidationDatum[]): any[];
declare function ensure(validationDatum: ValidationDatum, options: EnsureOptions): any[];
declare function ensure(validationDatum: ValidationDatum, [...validationDatum: ValidationDatum[], options: EnsureOptions]): any[];
```

Reply:
`...validationDatum: ValidationDatum, options?: EnsureOptions` yep, not possible like that as the `...` has to be the last argument.

Reply:
Ok, so the way you've configured it is the best we can do.